### PR TITLE
Add react remapping to import resolver plugin

### PR DESCRIPTION
**Pull request by becca-bailey**
This work is also incorporated in #37385, but this is just the module replacement that allows Bento components to use external libraries that rely on React. 

The original problem: `react-day-picker` (the library that is used for the date picker under the hood) relies on React. Without any module replacement, rendering the day picker component in Bento mode doesn't render the calendar and fails silently. Aliasing `react` to `preact/compat` also results in some weird compatibility issues between `preact/compat` and `src/preact`. 

While the compiled bento code `bento.js` or `bento.max.js` also exports all the necessary React functions from `src/preact`, it is incompatible with synthetic default import syntax (`import React from 'react'`) used by many libraries.

Doing this module replacement in the babel config makes the most sense for now, and works with both test files and bento components.

### CODE REVIEW

**Review by jridgewell — APPROVED** [review]

File: src/react/index.js [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+  useLayoutEffect,
+  useMemo,
+  useRef,
+  useState,
+} from '#preact';
+import {forwardRef} from '#preact/compat';
+
+// This file allows us to remap react imports from external libraries to
+// our internal preact exports. This file uses a default export in order to
+// be compatible with libraries that use react default import syntax.
```
Comment (scottrippey) [inline src/react/index.js:17]: Thank you for this explanation!

File: build-system/babel-config/import-resolver.js [hunk 2]
```diff
@@ -41,14 +41,23 @@ function readJsconfigPaths() {
   return tsConfigPaths;
 }
 
+// Necessary for external modules that rely on React.
+const moduleAliases = {
+  'react': './src/react',
```
Comment (scottrippey) [inline build-system/babel-config/import-resolver.js:46]: Ah, this is great ... the previous attempt was much more complicated, but this is so simple, by redirecting `react -> .../react.js -> #preact` we ensure that we always get the build config from the `#preact` alias.  Nice!

**Review by scottrippey — CHANGES_REQUESTED** [review]
Overall I love the simplicity of this change.

Unfortunately, it wasn't working correctly when I tried it locally.
In my branch (with this PR merged), when importing `react-hammerjs` it fails with this error message:
```
TypeError: Super expression must either be null or a function, not undefined
```

I'm tracing down the issue now, but just wanted to say something is still broken...

File: src/react/index.js [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+  createContext,
+  createElement,
+  useCallback,
+  useContext,
+  useEffect,
+  useImperativeHandle,
+  useLayoutEffect,
+  useMemo,
+  useRef,
+  useState,
```
Comment (scottrippey) [inline src/react/index.js:11]: Ok, I figured out my issue on my branch.  TLDR I had to add a lot of `import, export, import, export` for various features.

The good news is, after re-exporting all these features, it works!  
The bad news is, it took a lot of manual hunt-and-peck to find all the issues, and the process was pretty tedious.  
I'm going to experiment and see if I can do something along the lines of `export * from '#preact'`

Comment (scottrippey) [reply src/react/index.js:11]: This works for me: replace this whole file with:
```
import * as Preact from '#preact';
import * as PreactCompat from '#preact/compat';

export * from '#preact';
export * from '#preact/compat';

// This file allows us to remap react imports from external libraries to
// our internal preact exports. This file uses a default export in order to
// be compatible with libraries that use react default import syntax.
export default {...Preact, ...PreactCompat};
```

With this `*` in place, I only need to update `/src/preact/index.js` whenever a 3rd party lib tries to use a new API (eg. `useDebugValue` or `Component`)

Comment (becca-bailey) [reply src/react/index.js:11]: Sounds good, I will make this update!

**Review by dethstrobe — APPROVED** [review]

**Review by scottrippey — APPROVED** [review]

File: build-system/babel-config/import-resolver.js [hunk 4]
```diff
@@ -41,14 +41,33 @@ function readJsconfigPaths() {
   return tsConfigPaths;
 }
 
+/**
+ * Remap external modules that rely on React if building for Preact.
+ * @param {'preact' | 'react'} buildFor
+ * @return {Object}
+ */
+function moduleAliases(buildFor) {
+  if (buildFor == 'preact') {
```
Comment (samouri) [inline build-system/babel-config/import-resolver.js:50]: Would this behave any different if we always applied it vs. only applying for preact builds?

**Review by samouri — APPROVED** [review]
> Aliasing `react` to `preact/compat` also results in some weird compatibility issues between `preact/compat` and `src/preact`. 

Did you try aliasing `react` to `src/preact`?

---

Also, thank you for working through this!

**Comment by becca-bailey** [comment]
@samouri `src/preact` does not include a default export, which is required for libraries that use a default import syntax for React. `src/react` imports the same functions from `src/preact`, but uses a default export in order to be an alias for React.

**Comment by samouri** [comment]
> @samouri `src/preact` does not include a default export, which is required for libraries that use a default import syntax for React. `src/react` imports the same functions from `src/preact`, but uses a default export in order to be an alias for React.

Adding the default export line to `src/preact` sounds like an overall smaller diff. I think it would be overall cleaner if we eventually coalesced everything into either a `src/react` _or_ a `src/preact`

File: build-system/babel-config/import-resolver.js (hunk 4 above)
Comment (dethstrobe) [reply build-system/babel-config/import-resolver.js:50]: Good thinking. I'll make this is the default

**Comment by becca-bailey** [comment]
@samouri Yep, we can definitely continue to iterate! Right now there is a linter rule in place that bans default exports from `src/preact`, so it might be good to investigate a little more to figure out whether this smaller change would affect other parts of the build process.

**Pull request merged**